File: cpp/dsa/data-structures/binary-tree/diameter-of-binary-tree.cpp
/** Definition for a binary tree node */
// struct TreeNode { int val; TreeNode *left; TreeNode *right; }

class Solution {
    int diameter = 0;
public:
    int diameterOfBinaryTree(TreeNode* root) { // @viz: function_start
        height(root); // @viz: height_start
        return diameter; // @viz: function_start
    }
    
    int height(TreeNode* node) { // @viz: height_start
        if (!node) return 0; // @viz: base_case
        
        int left = height(node->left); // @viz: recurse_left
        int right = height(node->right); // @viz: recurse_right
        
        diameter = max(diameter, left + right); // @viz: update_diameter
        return 1 + max(left, right); // @viz: return_height
    }
};

File: python/dsa/data-structures/binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode: val, left, right

class Solution:
    def diameterOfBinaryTree(self, root): # @viz: function_start
        self.diameter = 0
        self.height(root) # @viz: height_start
        return self.diameter # @viz: function_start
    
    # Helper function to calculate height
    def height(self, node): # @viz: height_start
        if not node: return 0 # @viz: base_case
        
        left = self.height(node.left) # @viz: recurse_left
        right = self.height(node.right) # @viz: recurse_right
        
        self.diameter = max(self.diameter, left + right) # @viz: update_diameter
        return 1 + max(left, right) # @viz: return_height

